File: Infosys/61.rotate-list.java
/**
 * Algorithm:
 * Compute length, connect tail to head to form a cycle, then break the cycle at
 * the new tail. Rotating right by k makes the length - k % length node the new
 * head.
 *
 * Complexity:
 * Time O(n), space O(1).
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) {
            return head;
        }
        int length = 1;
        ListNode tail = head;
        while (tail.next != null) {
            tail = tail.next;
            length++;
        }
        k %= length;
        if (k == 0) {
            return head;
        }
        tail.next = head;
        int stepsToNewTail = length - k - 1;
        ListNode newTail = head;
        for (int i = 0; i < stepsToNewTail; i++) {
            newTail = newTail.next;
        }
        ListNode newHead = newTail.next;
        newTail.next = null;
        return newHead;
    }
}

File: Walmart Labs/61.rotate-list.py
#
# @lc app=leetcode id=61 lang=python3
#
# [61] Rotate List
#
# https://leetcode.com/problems/rotate-list/description/
#
# algorithms
# Medium (42.52%)
# Likes:    11454
# Dislikes: 1558
# Total Accepted:    1.7M
# Total Submissions: 4M
# Testcase Example:  '[1,2,3,4,5]\n2'
#
# Given the head of a linked list, rotate the list to the right by k places.
# 
# 
# Example 1:
# 
# 
# Input: head = [1,2,3,4,5], k = 2
# Output: [4,5,1,2,3]
# 
# 
# Example 2:
# 
# 
# Input: head = [0,1,2], k = 4
# Output: [2,0,1]
# 
# 
# 
# Constraints:
# 
# 
# The number of nodes in the list is in the range [0, 500].
# -100 <= Node.val <= 100
# 0 <= k <= 2 * 10^9
# 
# 
#

# @lc code=start
from typing import List, Optional
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        """
        Interview explanation:
        Rotating right by k makes the (length - k % length)-th node the new head.
        First compute length and temporarily connect tail to head, forming a
        cycle. Then break the cycle at the new tail. This avoids moving nodes one
        rotation at a time.

        Edge cases and tests:
        - Empty list and one-node list are unchanged.
        - k == 0 or k multiple of length returns original order.
        - k greater than length is reduced by modulo.

        Complexity: O(n) time, O(1) space.
        """
        if not head or not head.next or k == 0:
            return head

        length = 1
        tail = head
        while tail.next:
            tail = tail.next
            length += 1

        k %= length
        if k == 0:
            return head

        tail.next = head
        steps_to_new_tail = length - k - 1
        new_tail = head
        for _ in range(steps_to_new_tail):
            new_tail = new_tail.next
        new_head = new_tail.next
        new_tail.next = None
        return new_head
# @lc code=end


